libkbfs/file_data.go

```diff
@@ -520,44 +543,294 @@ func (fd *fileData) createIndirectBlock(
+		lowestAncestorWithRoom)
+
+	// Make a new right block for every parent, starting with the
+	// lowest ancestor with room.  Note that we're not iterating over
+	// the actual parent blocks here; we're only using its length to
+	// figure out how many levels need new blocks.
+	pblock := parentBlocks[lowestAncestorWithRoom].pblock
+	for i := lowestAncestorWithRoom; i < len(parentBlocks); i++ {
+		newRID, err := fd.crypto.MakeTemporaryBlockID()
+		if err != nil {
+			return nil, nil, err
+		}
 
-	pblock.IPtrs = append(pblock.IPtrs, IndirectFilePtr{
-		BlockInfo: BlockInfo{
-			BlockPointer: newPtr,
-			EncodedSize:  0,
-		},
-		Off: off,
-	})
+		newPtr := BlockPointer{
+			ID:      newRID,
+			KeyGen:  fd.kmd.LatestKeyGeneration(),
+			DataVer: dver,
+			Context: kbfsblock.MakeFirstContext(fd.uid),
+		}
 
-	err = fd.cacher(newPtr, rblock)
-	if err != nil {
-		return err
+		fd.log.CDebugf(ctx, "New right block for file %d, level %d, ptr %v",
+			fd.rootBlockPointer(), i, newPtr)
+
+		pblock.IPtrs = append(pblock.IPtrs, IndirectFilePtr{
+			BlockInfo: BlockInfo{
+				BlockPointer: newPtr,
+				EncodedSize:  0,
+			},
+			Off: off,
+		})
+		rightParentBlocks[i].pblock = pblock
+		rightParentBlocks[i].childIndex = len(pblock.IPtrs) - 1
+
+		rblock := &FileBlock{}
+		if i != len(parentBlocks)-1 {
+			rblock.IsInd = true
+			pblock = rblock
+		}
+
+		err = fd.cacher(newPtr, rblock)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		newDirtyPtrs = append(newDirtyPtrs, newPtr)
 	}
-	err = fd.cacher(ptr, pblock)
-	if err != nil {
-		return err
+
+	// All parents up to and including the lowest ancestor with room
+	// will have to change, so mark them as dirty.
+	ptr := fd.rootBlockPointer()
+	for i := 0; i <= lowestAncestorWithRoom; i++ {
+		pb := parentBlocks[i]
+		if err := fd.cacher(ptr, pb.pblock); err != nil {
+			return nil, nil, err
+		}
+		newDirtyPtrs = append(newDirtyPtrs, ptr)
+		ptr = pb.childIPtr().BlockPointer
+		rightParentBlocks[i].pblock = pb.pblock
+		rightParentBlocks[i].childIndex = len(pb.pblock.IPtrs) - 1
 	}
-	return nil
+
+	return rightParentBlocks, newDirtyPtrs, nil
+}
+
+// shiftBlocksToFillHole should be called after newRightBlock when the
+// offset for the new block (`newHoleStartOff`) is smaller than the
+// size of the file.  This happens when there is a hole in the file,
+// and the user is now writing data into that hole.  This function
+// moves the new block into the correct place, and rearranges all the
+// indirect pointers in the file as needed.  It returns any block
+// pointers that were dirtied in the process.
+func (fd *fileData) shiftBlocksToFillHole(
+	ctx context.Context, newTopBlock *FileBlock, newHoleStartOff int64,
+	parents []parentBlockAndChildIndex) (
+	newDirtyPtrs []BlockPointer, err error) {
+	fd.log.CDebugf(ctx, "Shifting block with offset %d for file %v into "+
+		"position", newHoleStartOff, fd.rootBlockPointer())
+
+	// Walk down the right side of the tree to find the new rightmost
```

Comment:
The first part of this comment needs to be updated, since we're not walking down the right side anymore.